theano/tensor/var.py

```diff
@@ -460,30 +460,48 @@ def astype(self, dtype):
     # SLICING/INDEXING
     def __getitem__(self, args):
 
-        def check_bool(args_el):
-            try:
-                if (isinstance(args_el, (np.bool_, bool)) or
-                        args_el.dtype == 'bool'):
-                    raise TypeError('TensorType does not support boolean '
-                                    'mask for indexing such as tensor[x==0]. '
-                                    'Instead you can use non_zeros() such as '
-                                    'tensor[(x == 0).nonzeros()]. ')
-            except AttributeError:
-                pass
-
+        def includes_bool(args_el):
+            if (isinstance(args_el, (np.bool_, bool)) or
+                    (hasattr(args_el, 'dtype') and args_el.dtype == 'bool')):
+                return True
             if (not isinstance(args_el, theano.tensor.Variable) and
                     isinstance(args_el, collections.Iterable)):
                 for el in args_el:
-                    check_bool(el)
-
-        check_bool(args)
+                    if includes_bool(el):
+                        return True
+            return False
 
         if (isinstance(args, list) and
                 any([isinstance(a, slice) for a in args])):
             pass
         elif not isinstance(args, tuple):
             args = args,
 
+        # Convert boolean arrays to calls to mask.nonzero()
+        tmp_args = []
+        for arg in args:
+            # NumPy arrays or tensors of type bool can be converted to
+            # normal integer indices.
+            if (isinstance(arg, (np.ndarray, theano.tensor.Variable)) and
+                    hasattr(arg, 'dtype') and hasattr(arg, 'nonzero') and
+                    arg.dtype == 'bool'):
```

Comment:
That's what I tried first, but that fails on this test:
```python
    assert_array_equal(numpy_n[:, mask], n[:, self.shared(mask)].eval())
```
because `theano.tensor.sharedvar.TensorSharedVariable` is not a `TensorVariable`. There is a similar problem `TensorConstant`, `GpuArraySharedVariable` et cetera. This was the least ugly solution I could think of that covers all those types.